src/app/shared-space.tsx

```diff
@@ -1,131 +1,211 @@
-  
-
-        const svgPath = voronoi.render();
-        setSvgPath(svgPath)
-    }, [points, svgRef, windowDimensions])
-
-    if (!svgRef.current) return null
-
-    return (
-        <path d={svgPath} fill="none" stroke="black" />
-    )
+"use client";
+
+import { useEffect, useState, useRef } from "react";
+
+import { useCursors } from "./cursors-context";
+import OtherCursor from "./other-cursor";
+import SelfCursor from "./self-cursor";
+
+import * as d3 from "d3";
+import { Delaunay } from "d3-delaunay";
+
+type Point = [number, number, string]; // [x, y, id]
+
+function Voronoi(props: {
+  svgRef: React.RefObject<SVGSVGElement>;
+  windowDimensions: { width: number; height: number };
+}) {
+  const { others, self } = useCursors();
+  const [points, setPoints] = useState<Point[]>([]);
+  const { svgRef, windowDimensions } = props;
+  const [svgPath, setSvgPath] = useState<string>("");
+
+  useEffect(() => {
+    // If there aren't any other points, put a virtual point in the center of the window so the user has something to play with
+    let points = Object.entries(others).map(
+      ([id, { x, y }]) =>
+        [x * windowDimensions.width, y * windowDimensions.height, id] as Point
+    );
+    if (self) {
+      points.push([
+        self.x * windowDimensions.width,
+        self.y * windowDimensions.height,
+        "self",
+      ] as Point);
+    }
+    if (points.length <= 1) {
+      points.push([
+        windowDimensions.width / 2,
+        windowDimensions.height / 2,
+        "dummy",
+      ] as Point);
+    }
+    setPoints(points);
+  }, [others, self, windowDimensions]);
+
+  useEffect(() => {
+    const svg = d3.select(svgRef.current);
+    svg
+      .attr("width", windowDimensions.width)
+      .attr("height", windowDimensions.height);
+    svg.attr(
+      "viewBox",
+      `0 0 ${windowDimensions.width} ${windowDimensions.height}`
+    );
+
+    // @ts-ignore
+    const delaunay = Delaunay.from(points);
+    const voronoi = delaunay.voronoi([
+      0,
+      0,
+      windowDimensions.width,
+      windowDimensions.height,
+    ]);
+
+    const pattern = (i: number) => {
+      return i % 2 === 0
+        ? "url(#diagonal-stripe-1)"
+        : "url(#diagonal-stripe-4)";
+    };
+
+    function stringToHash(str: string): number {
+      let hash = 0;
+      for (let i = 0; i < str.length; i++) {
+        const char = str.charCodeAt(i);
+        hash = (hash << 5) - hash + char;
+        hash |= 0; // Convert to 32bit integer
+      }
+      return hash;
+    }
+
+    // Go from a connection string id to a pastel color
+    // The trick is to work in HSL, fixing S and L, and using i to pick a hue
+    // Then convert back to RGB hex
+    const pastel = (id: string) => {
+      const hash = stringToHash(id);
+      // OFFSET is a magic number. Adjust this such that the colour for the zeroth connection
+      // (called "dummy" above) and first human connection is pleasing
+      const OFFSET = 7;
```

Comment:
Is this offset still needed? I believe we added this just so the first connecting colour would be purplish instead of reddish to make the single-user visual experience a little more palatable, but now with id hashing, the first colour is no longer deterministic, so this is not doing what it used to do anymore.